QUALITY_DECLARATION.md

```diff
@@ -0,0 +1,132 @@
+
+### Copyright Statements
+
+The copyright holders each provide a statement of copyright in each source code file in `rcutils`.
+
+There is an automated test which runs a linter that ensures each file has at least one copyright statement.
+
+## Testing
+
+### Feature Testing
+
+Each feature in `rcutils` has corresponding tests which simulate typical usage, and they are located in the `test` directory.
+New features are required to have tests before being added.
+
+### Public API Testing
+
+Each part of the public API have tests, and new additions or changes to the public API require tests before being added.
+The tests aim to cover both typical usage and corner cases, but are quantified by contributing to code coverage.
+
+### Coverage
+
+TODO fix link
+
+`rcutils` follows the recommendations for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#coverage), and opts to use branch coverage instead of line coverage.
+
+This includes:
+
+- tracking and reporting branch coverage statistics
+- achieving and maintaining branch coverage at or above 95%
+- no lines are manually skipped in coverage calculations
+
+Changes are required to make a best effort to keep or increase coverage before being accepted, but decreases are allowed if properly justified and accepted by reviewers.
+
+Current coverage statistics can be viewed here:
+
+TODO FIXME
+
+### Performance
+
+TODO fix link
+
+`rcutils` follows the recommendations for performance testing of C code in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#performance-c), and opts to do performance analysis on each release rather than each change.
+
+TODO how to run perf tests, where do they live, etc.
+
+TODO exclusions of parts of the code from perf testing listed here? or linked to?
+
+### Linters and Static Analysis
+
+TODO fix link
+
+`rcutils` uses and passes all the standard linters and static analysis tools for a C package as described in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#linters-and-static-analysis).
+
+TODO any qualifications on what "passing" means for certain linters
+
+## Dependencies
+
+`rcutils` has no run-time or build-time dependencies that need to be considered for this declaration.
+
+It has several "buildtool" dependencies, which do not affect the resulting quality of the package, because they do not contribute to the public library API.
+It also has several test dependencies, which do not affect the resulting quality of the package, because they are only used to build and run the test code.
+
+## Platform Support
+
+`rcutils` supports all of the tier 1 platforms as described in [REP-2000](https://www.ros.org/reps/rep-2000.html#support-tiers), and tests each change against all of them.
+
+TODO make additional statements about non-tier 1 platforms?
```

Comment:
It would also be nice to add a link to supporting evidence here. CI results? Binary packages?

Reply:
I think for this would be a good idea to have a link pointing to the latest success jobs in packaging for each of the tier1 platforms. Like the panel shown on the main page of [CI](https://ci.ros2.org/) but only showing some metrics, maybe including test metrics here too. Not sure if this can be generated easily.

Reply:
Each job has a badge already, we can use those from the packaging jobs, like these:

- https://ci.ros2.org/view/packaging/job/packaging_linux/ [![Build Status](https://ci.ros2.org/view/packaging/job/packaging_linux/badge/icon)](https://ci.ros2.org/view/packaging/job/packaging_linux/)
- https://ci.ros2.org/view/packaging/job/packaging_windows/ [![Build Status](https://ci.ros2.org/view/packaging/job/packaging_windows/badge/icon)](https://ci.ros2.org/view/packaging/job/packaging_windows/)

But those are for all of the things in the `ros2.repos` file, not just this package...

We also have the `.deb` packaging jobs, which are per package, but they are only for Linux and aren't built nightly (right now), e.g.:

- http://build.ros2.org/view/Ebin_uB64/job/Ebin_uB64__rcutils__ubuntu_bionic_amd64__binary/ [![Build Status](http://build.ros2.org/view/Ebin_uB64/job/Ebin_uB64__rcutils__ubuntu_bionic_amd64__binary/badge/icon)](http://build.ros2.org/view/Ebin_uB64/job/Ebin_uB64__rcutils__ubuntu_bionic_amd64__binary/)

There are dev jobs, but those are per repository, not per package, e.g.:

- http://build.ros2.org/job/Edev__rcutils__ubuntu_bionic_amd64/ [![Build Status](http://build.ros2.org/job/Edev__rcutils__ubuntu_bionic_amd64/badge/icon)](http://build.ros2.org/job/Edev__rcutils__ubuntu_bionic_amd64/)

Reply:
Yeah, I thought about those, in the sense that these links can be referenced and checked easily, but done for each package. 

Maybe we could point to the latest successful build tests of each package, per platform, something like what's found here:
[Test results ament_index_python nightly_osx_debug](https://ci.ros2.org/job/nightly_osx_debug/lastSuccessfulBuild/testReport/ament_index_python.test/).